File: python3/easy-solution/20.py
"""
20. Valid Parentheses
    Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
    An input string is valid if:
    Open brackets must be closed by the same type of brackets.
    Open brackets must be closed in the correct order.
    Every close bracket has a corresponding open bracket of the same type.
        
    Example:
    Input: s = "()"
    Output: true
"""

class Solution:
    def isValid(self, s: str) -> bool:
            sym = []
            if(s[0]==')' or s[0]==']' or s[0]=='}'):
                return False
            for i in s:
                if (i=='(' or i=='[' or i=='{'):
                    sym.append(i)
                elif(len(sym)!= 0 and (sym[-1]=='(' or sym[-1]=='[' or sym[-1]=='{')):
                    if((i==')' and sym[-1]=='(') or (i==']' and sym[-1]=='[') or (i=='}' and sym[-1]=='{')):
                        sym.pop()
                    else:
                        return False
                else:
                    return False
            if(len(sym)==0):
                return True
            return False

File: cpp/easy-solution/20.cc
/**
 * 20. Valid Parentheses
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * An input string is valid if:
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Every close bracket has a corresponding open bracket of the same type.
 * 
 * Example:
 * Input: s = "()"
 * Output: true
 */

#include <stack>
#include <string>

class Solution {
public:
    bool isValid(std::string s) {
        std::stack<char> sym;
        if (s[0] == ')' || s[0] == ']' || s[0] == '}') {
            return false;
        }
        for (char i : s) {
            if (i == '(' || i == '[' || i == '{') {
                sym.push(i);
            } 
            else if (!sym.empty() && (sym.top() == '(' || sym.top() == '[' || sym.top() == '{')) {
                if ((i == ')' && sym.top() == '(') || (i == ']' && sym.top() == '[') || (i == '}' && sym.top() == '{')) {
                    sym.pop(); 
                } else {
                    return false; 
                }
            } else {
                return false; 
            }
        }
        return sym.empty();
    }
};